### DISCUSSION TITLE
Please provide an AppImage for Linux

### DISCUSSION THREAD

**Discussion by probonopd** [mu-editor/mu#1402] [Ideas & New Features]
Currently it is extremely easy to download and run Mu on the Mac, but on Linux the command line is involved, making it more complex than on the Mac.

Providing an [AppImage](http://appimage.org/) would have, among others, these advantages:
- Applications packaged as an AppImage can run on many distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, [make it executable](http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80), and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Works out of the box, no installation of runtimes needed
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs
- Can use the same AppImages when dual-booting multiple distributions
- Can be listed in the [AppImageHub](https://appimage.github.io/apps) central directory of available AppImages
- Can double as a self-extracting compressed archive with the `--appimage-extract` parameter

An AppImage is essentially a self-mounting filesystem (similar to a Mac disk image) that contains Python, PyQt, Python libraries, icons, other resources, and the application.

Complex Python/PyQt applications such as Ultimaker Cura are already being shipped in AppImage format.

[Here is an overview](https://appimage.github.io/apps) of projects that are already distributing upstream-provided, official AppImages.

Here is a simple recipe that can produce an AppImage from pip:
https://github.com/AppImage/AppImages/blob/master/recipes/Mu.yml

On a Debian/Ubuntu based distribution, you can run it with

```
wget https://raw.githubusercontent.com/AppImage/AppImages/master/pkg2appimage
bash -ex pkg2appimage Mu
```

If you have questions, AppImage developers are on #AppImage on irc.freenode.net.

**Comment by ntoll** [comment]
Oooh... I didn't know about this. I'll take a look.

A quick question: what testing is available to check the appimage on all supported platforms?

**Comment by probonopd** [comment]
I am using [`testappimage`](https://github.com/AppImage/AppImageKit/blob/appimagetool/master/testappimage) along with Live ISOs of various distros.

**Comment by abixadamj** [comment]
I agree that AppImage would be great solution! I will help with creating this.

**Comment by probonopd** [comment]
On a Debian/Ubuntu based distribution, please try

```
wget https://raw.githubusercontent.com/AppImage/AppImages/master/pkg2appimage
bash -ex pkg2appimage Mu
```

It should produce a working AppImage.

**Comment by tim1mw** [comment]
+1 for this, I would have much preferred an appimage.

 I'm currently wrestling with the pip install which is crashing due to other changes in the latest distros (#875 )  so I can't currently run mu at all.

**Comment by tmontes** [comment]
All,

Letting you know that I have made substantial progress towards the goal packaging Mu Editor as a Linux AppImage.

Details in this [`pup` issue](https://github.com/mu-editor/pup/issues/150), in this [just merged `pup` PR](https://github.com/mu-editor/pup/pull/183), and in this [Mu Editor PR](https://github.com/mu-editor/mu/pull/1863).

Once these items are ironed out we should be able to ship our first "official" Mu Editor AppImages -- whether they will work and make our Linux users' lives easier or not is something that we'll have to wait for and see. :)

Stay tuned!

**Reply by probonopd** [reply]
Cool! Please let me know when there is an AppImage that can be tested.

**Comment by tmontes** [comment]
All,

First attempt at AppImage Linux packaging for the current release of Mu just automatically built available [at the bottom of this page](https://github.com/mu-editor/mu/actions/runs/2024827478).

Just tested it on Debian 10:
* Download.
* Make executable.
* Run it!

Let me know how it goes, please! :)

**Reply by probonopd** [reply]
Thanks @tmontes. Right now I am getting `qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.` (full disclaimer: on [helloSystem](https://hellosystem.github.io/) FreeBSD with a Debian runtime), so I am wondering what tools you are using to bundle the required dependencies of Qt.

**Reply by carlosperate** [reply]
I gave this a quick test in a Linux Mint 19.1 VM (fairly out of the box clean install, nothing apt installed yet), and it worked! 🎉

> `qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.` (full disclaimer: on [helloSystem](https://hellosystem.github.io/) FreeBSD with a Debian runtime)

Is the issue similar to this? https://github.com/mu-editor/mu/pull/1272#issuecomment-779130848

If so, are you able to check what version of libxcb-util you might have (if any)?

@tmontes I'm a bit surprised to see this one, when I encountered it, it was in Ubuntu 20.04, and this is build in 18.04, which should come with libxcb-util.so.0. I guess this lib is not being packaged? Is there a way to check what is AppImage collecting?

**Reply by tmontes** [reply]
AppImage packaged Mu does not ship with Qt dependencies that something akin to `pip install PyQt` requires from the underlying system.

From a cursory glance -- after extracting the AppImage with `--appimage-extract` there are many many Qt shared libraries that depend on the following "xcb"-like existing libraries on the system:

```
libX11-xcb.so.1
libxcb-glx.so.0
libxcb-render.so.0
libxcb-shm.so.0
libxcb.so.0
libxcb-xkb.so.0
```

PS: Hand copied, so there might be minor mistakes!

On the debian 10 system I tried, there are no less than "xcb"-like system packages installed. I suppose they are providing the shared libs listed above.

Maybe the system you are running @probonopd does not have them... Can you confirm that? Can you figure a way of installing the FreeBSD package(s) that bring them in, if needed?

Thanks for your assistance.

More PS: This is admitedly a hard-core, non-common case, I'd say... FreeBSD? helloSystem? Cool! :)

**Reply by tmontes** [reply]
Played with it a little bit more, under debian 10... Then explored and found https://github.com/helloSystem/QtPlugin. Could this be related? (no idea!) :)

**Reply by tmontes** [reply]
At least, at some point in instructs FreeBSD users to
```
sudo pkg install -y curl zip cmake pkgconf libfm qt5-core qt5-x11extras qt5-widgets qt5-qmake qt5-buildtools qt5-linguisttools kf5-extra-cmake-modules libqtxdg libxcb libdbusmenu-qt5 kf5-kwindowsystem
```

...which seems to install **libxcb**, maybe the culprit. Could it be that installing that package on your systems makes Mu work?

**Reply by probonopd** [reply]
> This is admitedly a hard-core, non-common case

Indeed. Hence this is more a "fyi" than anything.

> Maybe the system you are running @probonopd does not have them... Can you confirm that? 

[Here](https://github.com/helloSystem/LinuxRuntime/blob/740506832241c1e8f6711cd11988e95363eb72aa/prepare-debian-userland.sh#L19-L23) are the libraries I have installed.

**Reply by probonopd** [reply]
> AppImage packaged Mu does not ship with Qt dependencies that something akin to pip install PyQt requires from the underlying system.

It is recommended to privately bundle all libraries that Qt depends on, with the possible exception of those explicitly listed on https://github.com/AppImage/pkg2appimage/blob/master/excludelist. Tools like [linuxdeployqt](https://github.com/probonopd/linuxdeployqt), [linuxdeploy-plugin-qt](https://github.com/linuxdeploy/linuxdeploy-plugin-qt) and many [spezialized deploypent tools for Python](https://project-awesome.org/AppImage/awesome-appimage#deployment-tools-for-python-applications) usually do that.

**Reply by probonopd** [reply]
@tmontes [QtPlugin](https://github.com/helloSystem/QtPlugin) is completely unrelated. That is the FreeBSD side of things. In oder to support Linux AppImages, helloSystem is using a dedicated [Debian Runtime](https://github.com/helloSystem/LinuxRuntime/).

**Reply by carlosperate** [reply]
From that AppImage exclude list:

```
libxcb.so.1
# Workaround for:
# Fedora 23
# symbol lookup error: /lib64/libxcb-dri3.so.0: undefined symbol: xcb_send_fd
# Uncertain if this is required to be bundled for some distributions - if so we need to write a version check script and use LD_PRELOAD to load the system version if it is newer
# Fedora 25:
# undefined symbol: xcb_send_request_with_fds
# https://github.com/AppImage/AppImages/issues/128
```

Could that be relevant?

**Reply by probonopd** [reply]
I don't think so but let's not invest too much time into my "unusual" setup for now. Once there is a downloadable AppImage (not inside a zip) I can run it through the [automated AppImageHub test](https://github.com/AppImage/appimage.github.io/).

**Reply by carlosperate** [reply]
@probonopd there is downloadable AppImage with this URL in case you'd like to run it through the automated test: https://github.com/mu-editor/mu/releases/download/v1.1.1/Mu_Editor-1.1.1-x86_64.AppImage

**Comment by abixadamj** [comment]
Just for a record. Working fine on Ubuntu 20.04, but I have Mu in local got cloned repo, so I have all neccessary libraries in system.
![Zrzut ekranu z 2022-03-25 16-14-59](https://user-images.githubusercontent.com/5088643/160148815-9dbaf61a-294e-4f6e-9105-c28523508814.png)

**Reply by carlosperate** [reply]
Thanks for testing!
I also tried it on a brand new VM install of 20.04 and it worked well 🎉

**Comment by dhalbert** [comment]
Tested on Ubuntu 22.04 x64 (arriving next month -- note 22.04, not 20.04).
```
$ Downloads/Mu_Editor-x86_64.AppImage 
dlopen(): error loading libfuse.so.2

AppImages require FUSE to run. 
You might still be able to extract the contents of this AppImage 
if you run it with the --appimage-extract option. 
See https://github.com/AppImage/AppImageKit/wiki/FUSE 
for more information
```
22.04 has `libfuse3`. I installed libfuse2, but no joy:
```
$ Downloads/Mu_Editor-x86_64.AppImage 
No settings file found at /home/halbert/.local/share/mu/venv.json; skipping
Segmentation fault (core dumped)
```

**Reply by carlosperate** [reply]
@tmontes I can confirm that on a fresh install of Ubuntu 22.04 we get that FUSE error message, and after installing `libfuse2` there is a seg fault when trying to run the AppImage.

```
$ ./Mu_Editor-1.1.1-x86_64.AppImage 
No settings file found at /home/carlos/.local/share/mu/venv.json; skipping
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Segmentation fault (core dumped)
```

**Reply by carlosperate** [reply]
The seg fault might be an X11 issue, as this seems to work:

```
$ QT_QPA_PLATFORM=wayland ./Mu_Editor-1.1.1-x86_64.AppImage 
```

And these segfaults:
```
$ QT_QPA_PLATFORM=xcb ./Mu_Editor-1.1.1-x86_64.AppImage 
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Segmentation fault (core dumped)
```
```
$ env -u WAYLAND_DISPLAY ./Mu_Editor-1.1.1-x86_64.AppImage 
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Segmentation fault (core dumped)
```

**Comment by ntoll** [comment]
Experimental AppImage version added to our downloads page. Please let us know how you get on. Thank you!

**Comment by tmontes** [comment]
All,

We've updated the docs on using the Linux AppImage at https://codewith.mu/en/howto/1.1/install_linux.
Planning to update as others try it out under other Linux distributions.

Your feedback will be welcome. Thanks!

**Reply by probonopd** [reply]
Thanks for making an AppImage, which almost passes the test, but not quite: https://github.com/AppImage/appimage.github.io/pull/2925

**Comment by MMachado05** [comment]
Out of curiosity, what was the goal with packaging the AppImage inside an archive `tar` file? It seems to me that simply providing the AppImage solely and directly from the download button would provide a much quicker setup process.

**Reply by probonopd** [reply]
Indeed: https://docs.appimage.org/packaging-guide/distribution.html#do-not-put-appimages-into-other-archives

**Reply by carlosperate** [reply]
@probonopd do you know how providing the AppImage in a tar file breaks `appimaged` (and similar things) if the user unpacks the file anyway? I would assume, once the `.AppImage` file is in the file system everything should work as expected?

The main reason is archived into a tar file is specifically to preserve permissions. This is an application for very novice users, so we cannot assume they know how to configure this kind of thing. Providing instructions on how to change this in every distro/destop environment is extra work and extra maintenenance (and when things change it can be more confusing that useful), and even when we have instructions about setting permissions, we still had users not following the instructions and reporting issues with executing the file (and that's only those that are advance enough, care enough, or know about how to reach us via GitHub, gitter or other methods).

**Reply by probonopd** [reply]
@carlosperate it is an extra step to unpack the application from the archive, which (depending on the size of the application) can take additional time. Once the AppImage is unpacked, it works properly with `appimaged` of course. But you have this extra step rather than it being automatic.

**Reply by carlosperate** [reply]
Thanks for the quick response @probonopd! 

In my own personal opinion, I think the extra step is worth it. In all Linux systems I've tested, unpacking a tar file can be done via double clicking on the icon (with or without a GUI window with additional options), and the majority of users would be familiar with that operation already.
It might be a bit annoying having to wait for the extra step to unpack the tar file, but it's not even compressed, so hopefully it's not that bad, and it's a one-time-only step.

I would completely agree that for other applications, targetting more experience users, it can make sense to delivery the .AppImage file directly, and those users might be able to easily make the file executable. But one of the goals of Mu is to enable very novice users, on their first coding steps, who might not know much about how Linux/OSes (or even computers) work, and who more-often-than-not do not read instructions/documentation. So in those cases, friction at the beggining of their journey can impact quite negatively their motivation, and being able to double click on the .tar file + double click on the .AppImage file directly, that can make a huge difference.

**Reply by probonopd** [reply]
@carlosperate I see. Maybe it'd be a compomise to provide both an archived an an unarchived download?

**Reply by carlosperate** [reply]
Yes, I think that should be fine.
We upload the installers to each GitHub Release, and then link to them from the website. We could link to the archive file from the https://codewith.mu download page, and still have the .AppImage available in the GitHub release, so other people can download it from there. And I guess if we were to add it to the AppImage catalog, that would link to the direct file as well.
